### ISSUE TITLE
[$1000] Request Money – The address input lacks a red outline or error message when an incomplete IOU is created offline

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#32405]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.7.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in
3. Global Create > Request Money
4. Select Distance option
5. Disable the internet connection
6. Enter a vague start location like "SF"
7. Enter a vague finish location like "LA"
8. Finish the request creation flow
9. Enable the internet connection
10. Verify there's a red dot on the request report in the LHN, on the report preview in the workspace chat, and on the request
11. Navigate to the request with the error
12. Click on the request
13. Click on the start waypoint

## Expected Result:
The address input has a red outline and below it is an error message like “Unable to find the location for this address. Please edit the address and try again

## Actual Result:
The address input lacks a red outline or error message when an incomplete IOU is created offline

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/28d2dd28-3998-4a55-9c3a-ab91b6a28d04

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bc96448abc5e7e68</li>
        <li>Upwork Job ID: 1730973988318359552</li>
        <li>Last Price Increase: 2024-02-05</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Contributor | 28064604</li>
</ul></li>
    </ul>
</details>

**Comment by chiItepin** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The user can't tell if the **address** finder is **offline** due to a lack of a proper `hint` and inline validation, thus, causing the user to select an invalid address while being offline and being able to submit without proper validation such as:

<img width="371" alt="Screenshot 2023-12-02 at 18 12 25" src="https://github.com/Expensify/App/assets/50275849/75b60bff-f429-485e-9f92-fd5a548ce21c">


### What is the root cause of that problem?
There is no proper validation to tackle the lack of network connection in `WaypointEditor.js`, the form's validation excludes an the `isOffline` check as shown below:

```javascript
const validate = (values) => {
...
        // If the user is online, and they are trying to save a value without using the autocomplete, show an error message instructing them to use a selected address instead.
        // That enables us to save the address with coordinates when it is selected
        if (!isOffline && waypointValue !== '' && waypointAddress !== waypointValue) {
            ErrorUtils.addErrorMessage(errors, `waypoint${waypointIndex}`, 'distance.errors.selectSuggestedAddress');
        }
``` 


### What changes do you think we should make in order to solve the problem?

Ideally, a proper validation message should be shown to the user to let them know they are offline at the moment **if they are offline and an autocomplete item has not been selected** as shown below:

<img width="371" alt="Screenshot 2023-12-02 at 18 24 17" src="https://github.com/Expensify/App/assets/50275849/3488d9c5-6ab2-4ac0-a0ac-f05580523c12">

### What alternative solutions did you explore? (Optional)
A quick alternative would be to simply remove the `!isOffline` check from the `validate` function  at L123

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by conorpendergrast** [comment]
Reproduced on iOS (SafarI). @ntdiary two proposals for your review above!

**Comment by shubham1206agra** [comment]
Maybe taking over

**Comment by shubham1206agra** [comment]
@conorpendergrast Do we also need to show a red brick indicator in the Distance field on the request details page?

**Comment by situchan** [comment]
> @conorpendergrast Do we also need to show a red brick indicator in the Distance field on the request details page?

Being handled in #29783

**Comment by shubham1206agra** [comment]
@tienifr Can you point to the code where `pendingFields` was added?

**Comment by shubham1206agra** [comment]
> TBH, I am not sure this is worth fixing after #29783 is done, as we will show red dot already on 2 pages

I think we should have some indicator of which address has a problem. So that we can fix the address.

**Comment by chiItepin** [comment]
@shubham1206agra I see, but how is the user able to look for an address if the user is offline *and* there are no autocomplete items (recently added addresses)?

**Comment by shubham1206agra** [comment]
> @shubham1206agra I see, but how is the user able to look for an address if the user is offline _and_ there are no autocomplete items (recently added addresses)?

By manually entering the address and proceed with that

**Comment by chiItepin** [comment]
Hmm, I thought the app needed to sanitize the address, without being online and without autocomplete items, the user could literally enter anything thus, the map is not recognizing the address if this is the case which is what the reporter described though, map not being picked up due to invalid address, no?

**Comment by shubham1206agra** [comment]
We clear `pendingFields` when editing distance requests, too.
https://github.com/Expensify/App/blob/3827a9ce47f12f7c004339281c34a76f3a4c286b/src/libs/actions/IOU.js#L760

**Comment by shubham1206agra** [comment]
> @shubham1206agra we don't need to call API editDistanceMoney, just change the waypoint without clicking save button.

I didn't understood this comment @tienifr

**Comment by shubham1206agra** [comment]
@tienifr Maybe create pendingFields.route for that instance. If we run getRoute clear this field of pendingFields

**Comment by shubham1206agra** [comment]
> > If we run getRoute clear this field of pendingFields
> 
> You mean: clearing `pendingFields` even in optimistic data of `GetRoute` API?

No
What does getRoute return exactly? Can you give the Onyx update in response in both online and offline behaviour?

**Comment by shubham1206agra** [comment]
@tienifr I think it's not wise to modify the logic for `pendingFields`. I think we can use a field such as `isLoadingRoute` and set it accordingly, and use this instead.
Let me know your thoughts.

**Comment by shubham1206agra** [comment]
Can you update your proposal in that case?

**Comment by shubham1206agra** [comment]
@tienifr Please recheck your proposal. There seems to be changes in Money Request flow

**Comment by shubham1206agra** [comment]
Agreed. The flow recently changed. Let me check this again, and will update you here

**Comment by shubham1206agra** [comment]
@tienifr We are not using Form component in `WaypointEditor` (using FormProvider), and I am not in favour of using useEffect as it tends to get highly unstable in running order. Can you suggest another way for this?
And another PR is removing use of `pendingFields` in setting waypoint.

**Comment by shubham1206agra** [comment]
See https://github.com/Expensify/App/pull/33043

**Comment by conorpendergrast** [comment]
@shubham1206agra Sorry, I just realised that you weren't assigned as C+ here

**Comment by shubham1206agra** [comment]
> @shubham1206agra Thanks for pointing that out. Seems like we can't use pendingFields anymore. I just updated my proposal and added the new result video

@tienifr I have only one problem now, i.e., we do not want extra method of validation (only one validation method per form), and there can be prop provided to FormProvider to validate form on mounting.

**Comment by shubham1206agra** [comment]
@conorpendergrast What should we do here now? I don't like the current proposal is making changes to the Forms.

**Comment by shubham1206agra** [comment]
@conorpendergrast Bump on the above ^

**Comment by conorpendergrast** [comment]
@shubham1206agra Hey! I'm sorry, I was out of office. I think we can bump up the bounty as one method to attract more proposals. I'll set this back as external, and will do that now

**Comment by conorpendergrast** [comment]
I'm also checking the correct wave, [internally](https://expensify.slack.com/archives/C02MW39LT9N/p1704746319997709)

**Comment by dragnoir** [comment]
is this still open for proposals?

**Comment by conorpendergrast** [comment]
@dragnoir it is! Anything with Help Wanted and External is waiting for proposals

**Comment by Tony-MK** [comment]
Is anyone able to reproduce this bug?

This is my attempt on staging.

https://github.com/Expensify/App/assets/22982173/82808d02-02f3-4814-8822-c61d5cd2951f

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Request Money – The address input lacks a red outline or error message when an incomplete IOU is created offline

### What is the root cause of that problem?

We allow custom values to be inputted when offline

### What changes do you think we should make in order to solve the problem?

I think we have to prevent this from happening by validating the value even if it's offline. If we allow the custom text to be saved, it can lead to unexpected error. I've got "Hmm... page not found page" when re-open the preview.

We can change [this code](https://github.com/Expensify/App/blob/c14ca54536a50ecf251bd3ef7ebb4561f8e0e716/src/pages/iou/request/step/IOURequestStepWaypoint.js#L129-L133) to:

```
        // If the user is trying to save a value without using the autocomplete, show an error message instructing them to use a selected address instead.
        // That enables us to save the address with coordinates when it is selected
        if (waypointValue !== '' && waypointAddress !== waypointValue) {
            ErrorUtils.addErrorMessage(errors, `waypoint${pageIndex}`, 'distance.errors.selectSuggestedAddress');
        }

```

### What alternative solutions did you explore? (Optional)

**Comment by conorpendergrast** [comment]
@shubham1206agra Bump on the proposal review please!

**Comment by shubham1206agra** [comment]
@conorpendergrast I am OOO right now. I will review this on 30th

**Comment by conorpendergrast** [comment]
Sounds good, thanks!

**Comment by conorpendergrast** [comment]
It's not the 30th yet! That's tomorrow

**Comment by conorpendergrast** [comment]
Also [double-checking](https://expensify.slack.com/archives/C02MW39LT9N/p1706547680859439?thread_ts=1704746319.997709&cid=C02MW39LT9N) which wave this should be part of

**Comment by conorpendergrast** [comment]
@shubham1206agra can do, thanks for moving this along!

**Comment by conorpendergrast** [comment]
Ok, well, good news! Yes, it is expected that you can create a request like this. Quoting @neil-marcellini : 

> There’s no reasonable way to validate addresses while offline, but we want users to be able to enter addresses manually, or from their recent waypoints, and create distance requests while offline.

> However, we don’t handle errors for this flow well. It involves a mini project, described by this issue. [[DISTANCE] LOW: Handle errors with geocoding for offline distance requests](https://github.com/Expensify/App/issues/22719). I’m seeking someone to take it over

So, we'll close this out in favour of https://github.com/Expensify/App/issues/22719. That covers this scenario well! 

Thank you for your time everyone!

**Comment by shubham1206agra** [comment]
@conorpendergrast Can I get partial compensation in this case? As I have tried to move this along for a long time by doing quite back and forth with a proposal.

**Comment by conorpendergrast** [comment]
@shubham1206agra I can put that forward! What do you feel is a fair partial compensation?

**Comment by conorpendergrast** [comment]
[Posted internally](https://expensify.slack.com/archives/C01SKUP7QR0/p1707128213843969)

**Comment by conorpendergrast** [comment]
Still discussing internally

**Comment by conorpendergrast** [comment]
@shubham1206agra we've agreed with the 25% based on the work here. I've sent you [the offer](https://www.upwork.com/nx/wm/offer/100795410) now

**Comment by shubham1206agra** [comment]
@conorpendergrast Accepted

**Comment by conorpendergrast** [comment]
And paid! Thanks @shubham1206agra

**Comment by conorpendergrast** [comment]
Payouts due:

- [x] Contributor+: $250 @shubham1206agra  - [Contract](https://www.upwork.com/nx/wm/workroom/36272045/overview)

Upwork job is [here](https://www.upwork.com/jobs/~01bc96448abc5e7e68).
